### ISSUE TITLE
Offline isn't really offline anymore

### ISSUE THREAD

**Issue by ericzolf** [cgeo/cgeo#11725]
### Describe your problem!

Since the last major revamp of c:geo, even if solely using offline caches, C:geo still does something in the background. If the connection is bad, it will get stuck while loading the offline caches. The delay is gone if I go into airplane mode.

This didn't happen in the past, is cumbersome and potentially costs money.

### How to reproduce?

1. download offline caches
2. go to a place with bad connection
3. open the list of offline Caches

### Actual result after these steps?

Application hangs "forever"

### Expected result after these steps?

Application doesn't care and proceeds because it works *offline* by default.

### Reproducible

Yes

### c:geo Version

2021.09.08

### System information

```text
## System information

c:geo version: 2021.09.08

Device:- Device type: FP3 (FP3, Fairphone)
- Available processors: 8
- Android version: 10
- Android build: 8901.3.A.0129.20210805
- Screen resolution: 1080x2016px (360x648dp)
- Sailfish OS detected: false
- Google Play services: enabled - 21.33.14 (120408-395723304)
- HW acceleration: enabled (default state)

Sensor and location:- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: present
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:- Settings: v8, Count:136
- Set language: en_DE (system default)
- System date format: dd/MM/y
- Time zone: GMT+02:00
- Debug mode active: no
- Last backup: never
- Routing mode: Walk
- Live map mode: false
- OSM multi-threading: false / threads: 1
- Map: Germany__baden-wuerttemberg (Offline)
- Id: cgeo.geocaching.maps.mapsforge.MapsforgeMapProvider$OfflineMapSource:primary:Android/media/com.nextcloud.client/nextcloud/eric@oblok.lavarde.eu/mapsForge/germany__baden-wuerttemberg.map
- Atts: Map data (c) OpenStreetMap contributors
- Theme: none

Filters:- Hide waypoints: -
- LIVE: Difficulty / Terrain:1.0-3.0 / 1.0-3.0 ([:inconclusive=false:advanced=false]AND(type;difficulty_terrain:d=1.0:d=3.0:t=1.0:t=3.0;status))
- OFFLINE: None ([:inconclusive=false:advanced=false]AND(type;difficulty_terrain:d=-:d=-:t=-:t=-;status))

Services:- Geocaching sites enabled:
geocaching.com: Logged in (Login OK) / PREMIUM
- Geocaching.com date format: yyyy-MM-dd
- Routing: external / BRouter installed: false
- Installed c:geo plugins: none

Permissions & paths:- Fine location permission: granted
- Write external storage permission: granted
- System internal c:geo dir: /data/user/0/cgeo.geocaching (27,2 GB free) v2 internal isDir(7 entries)
- Legacy User storage c:geo dir: /storage/emulated/0/cgeo (27,2 GB free) v2 external non-removable isDir(7 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (27,2 GB free) v2 external non-removable isDir(876 entries)
- Internal theme sync (is turned off): /data/user/0/cgeo.geocaching/MapThemeData (27,2 GB free) v2 internal isDir(0 entries)
- Public Folders: #11
- BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=1, dirs:>=7, totalFileSize:>=3,9 KB, free space: 27,2 GB, files on device: 3194880)
- OFFLINE_MAPS: /Android/media/com.nextcloud.client/nextcloud/eric@oblok.lavarde.eu/mapsForge (User-Defined)[/Android/media/com.nextcloud.client/nextcloud/eric@oblok.lavarde.eu/mapsForge[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fmedia%2Fcom.nextcloud.client%2Fnextcloud%2Feric%40oblok.lavarde.eu%2FmapsForge::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fmedia%2Fcom.nextcloud.client%2Fnextcloud%2Feric%40oblok.lavarde.eu%2FmapsForge/document/primary%3AAndroid%2Fmedia%2Fcom.nextcloud.client%2Fnextcloud%2Feric%40oblok.lavarde.eu%2FmapsForge, Av:true, files:14, dirs:1, totalFileSize:1,6 GB, free space: 27,2 GB, files on device: 3194880)
- OFFLINE_MAP_THEMES: /Android/media/com.nextcloud.client/nextcloud/eric@oblok.lavarde.eu/mapsForge/_themes (Default)[/Android/media/com.nextcloud.client/nextcloud/eric@oblok.lavarde.eu/mapsForge/_themes[PERSISTABLE_FOLDER(OFFLINE_MAPS)#1:p-content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fmedia%2Fcom.nextcloud.client%2Fnextcloud%2Feric%40oblok.lavarde.eu%2FmapsForge::/_themes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fmedia%2Fcom.nextcloud.client%2Fnextcloud%2Feric%40oblok.lavarde.eu%2FmapsForge/document/primary%3AAndroid%2Fmedia%2Fcom.nextcloud.client%2Fnextcloud%2Feric%40oblok.lavarde.eu%2FmapsForge%2F_themes, Av:true, files:1, dirs:0, totalFileSize:0 B, free space: 27,2 GB, files on device: 3194880)
- LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:1, dirs:0, totalFileSize:3,9 KB, free space: 27,2 GB, files on device: 3194880)
- GPX: /cgeo/gpx (Default)[/cgeo/gpx[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/gpx]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fgpx, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 27,2 GB, files on device: 3194880)
- BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 27,2 GB, files on device: 3194880)
- FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 27,2 GB, files on device: 3194880)
- SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 27,2 GB, files on device: 3194880)
- ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:7, dirs:1, totalFileSize:81,2 KB, free space: 27,2 GB, files on device: 3194880)
- ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 27,2 GB, files on device: 3194880)
- TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 27,2 GB, files on device: -1)
- Map render theme path:
- PersistedDocumentUris: #1
- TRACK: null
- Persisted Uri Permissions: #2
- content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fmedia%2Fcom.nextcloud.client%2Fnextcloud%2Feric%40oblok.lavarde.eu%2FmapsForge (20 Apr, 08:43):RW
- content://com.android.externalstorage.documents/tree/primary%3Acgeo (20 Apr, 08:43):RW
- Database: /data/user/0/cgeo.geocaching/databases/data (v96, Size:27,1 MB) on system internal storage

--- End of system information ---
```

**Comment by Lineflyer** [comment]
Do you have the Geokrety service activated? (Asking due to #4779)

**Comment by ericzolf** [comment]
No, I don't (I didn't even know what it is, he ! ). I only have geocaching.com and GCVote active.

**Comment by Narrat** [comment]
I sometimes notice the same. Although not everytime when I have no connection.
Upon opening the cache details it hangs on "Loading Cache Details"

**Comment by moving-bits** [comment]
Can you cross-check whether this may be the cause?

c:geo will try downloading images for a cache listing, if they are not downloaded yet. So if you have imported a cache using a pocket query / gpx file, and open a listing containing images for the first time, there are no cached images yet.

**Comment by eddiemuc** [comment]
Afaik there is no explicit "offline mode" in c:geo anyway. E.g Cgeo will, if information is missing or related functions are triggered, always try to access information from the internet. Is an explicit "offline mode" requested as a feature?

**Comment by ericzolf** [comment]
> Can you cross-check whether this may be the cause?

All my cached caches are downloaded through the search function, so definitely not through pocket query / gpx, I know for sure because I have recently reinstalled c:geo.

> Is an explicit "offline mode" requested as a feature?

This wasn't my purpose but sounds also like an interesting idea (though I would rather ask for a "reduced data traffic on metered connection"), my point was more that, once caches are downloaded, c:geo shouldn't need to download anything more and hence shouldn't hang while waiting for data if the connection is bad.

**Comment by moving-bits** [comment]
> I just noticed that the option "Save Images" in the "Offline Data" settings is unselected. Do you think, this could be the issue?

Haven't tried, but may very well be the cause.

**Comment by SammysHP** [comment]
We got a report that might be related to this (or maybe not): https://www.geoclub.de/forum/t/listing-nicht-zu-%C3%B6ffnen.81879/

The user was unable to open a premium member cache (that was stored offline) after downgrading to a basic member. Later it worked again.

**Comment by Lineflyer** [comment]
We had recently user reports, that this happens "under bad network conditions".
While investigating I have seen one reproducible fact:
If you store a cache in c:geo and afterwards first time open it from the stored list, there is some network request involved. It looks like we are trying to fetch external content (pictures) embedded into the description.  On subsequent opening the cache no more online requests are made.
As I can't emulate bad network conditions I cannot tell whether this is a blocking action.

Is it possible we fail to store all pictures metadata while initially storing the cache, which leads to requesting them again?

**Comment by eddiemuc** [comment]
> If you store a cache in c:geo and afterwards first time open it [...] we are trying to fetch external content (pictures) embedded into the description. On subsequent opening the cache no more online requests are made.

Are the downloaded images maybe log images? Because those are not stored offline afaik (there's a setting for it).
 Htmlmage has some magic local cache for downloaded images, this could be the reason why images are downloaded only the first time.

**Comment by DerRadler** [comment]
Please add me to the list of interested parties hoping for a solution!  I am living in a region with wide areas of no mobile network (Germany <blush>).  I am bitten by this again and again.  Starting c:geo in flight mode to begin with does allow me to work with offline stored caches.  But as soon as I try to access an offline cache in connected mode with an nonworking connection, c:geo is blocked from accessing any offline cache forever.  Ten minutes, an hour, as long as it takes until I have a working network connection again.

I remember that in earlier versions of c:geo I had some timeout message after about a minute, when I tried to refresh a cache description while the network connection was not working.  Such a kind of timeout does not seem to be in effect now.  (I do not know who sponsored the timeout, c:geo itself or the OS API call).

After wrestling with this problem many times and after having lost several times, I can now at least work around it:
Close c:geo, then go to the Android application management and let it stop c:geo completely.  Then switch on flight mode.  And now loading c:geo again and access the offline stored cache listing.

Cheers, Frank der Radler

**Comment by Lineflyer** [comment]
I tried to find a clear clue here, but failed so far.
Currently I suspect, that it happens if a cache has linked resources (e.g. images) in listing which could not be stored offline while initially storing the cache (e.g. resource not available or not found). 
In such cases c:geo will try again to load the missing ressources but not sure though if this is a UI blocking problem.

@DerRadler 
If you happen to remember a GC code (or several) where that happened to you, it might be helpful for us to find a pattern/common reason causing this.

Does anyone know if the emulator is able to throttle incoming data to a low data rate?

**Comment by Lineflyer** [comment]
Addition...found something:
![image](https://user-images.githubusercontent.com/949669/164972788-544dad55-555b-4c7e-9094-91b61a03445f.png)

https://stackoverflow.com/questions/6236340/how-to-limit-speed-of-internet-connection-on-android-emulator

**Comment by DerRadler** [comment]
Lineflyer,
thank you for taking a crack at it!!

One GC Code I remember ... because it was driving me nuts ... is GC9JWJN.  I tried to access the offline description and then for the next hour was not able to access **any** other cache description as well.

I just tried to simulate the "in the woods" situation in our house.  Thought that because of the already weak signal it should be worse in the room of the central heating in the basement.  But there it still had connection.  I am not sure that it was just a slow connection.  Whenever I noticed that problem, I had no connection at all, but the phone was with mobile data enabled.

At nicer weather I will be happy to ride into the forest several times and try out any things you want me to try.  At the moment it is raining and I would rather stay inside.  If I can do any experiments with traces running (you would need to tell me how to start tracing), I will gladly do so.

**Comment by DerRadler** [comment]
There is one more fact to consider, which just came to my mind:
If we suspect that things need to be loaded by c:geo for certain caches, then maybe it does matter how the offline copy was stored.  In my case:  I am generating GPX files using GSAK.  These are synched to the phone.  Last, they are imported into c:geo by c:geo.  

I assume that these GSAK generated GPX are different to caches, which are searched through c:geo and then downloaded through c:geo.  If it would help I could provide a GSAK generated GPX file with any number of caches.

**Comment by ztNFny** [comment]
I also sometimes see this - but without a clear pattern so far. And as it only happens in the field I've been unable to reproduce it.
All my lists are refreshed inside c:geo though, so outdated ressources (which DerRadler suspected) should not be the cause.

**Comment by DerRadler** [comment]
Today I tried to test if I have this hang problem with a cache listing, which I had never looked at before and which I stored using c:geo.  In the forest with no mobile network coverage, c:geo handled that situation perfectly.  Told me that I had not been logged in and accessed the listing just fine. 

Next, I tried a listing, which I had brought into the phone as GSAK generated GPX file.  That failed on the first access but immediately went OK shortly after.  

So, in summary, this obnoxious hang, which I had three times on three different days, was not reproducible today.  Maybe this is really a matter of the network just failing at a certain time in a certain way?  Ob boy ....

**Comment by SammysHP** [comment]
If you have no connection at all then all network requests will fail immediately. If you have very bad connection, requests can be delayed for a long time (I've seen RTTs of 30 seconds and more) and will eventually aborted due to a timeout somewhere.

**Comment by DerRadler** [comment]
> If you have no connection at all then all network requests will fail immediately. If you have very bad connection, requests can be delayed for a long time (I've seen RTTs of 30 seconds and more) and will eventually aborted due to a timeout somewhere.

I thought so as well.  But in the past it has happened three times to me that there was no timeout.  I know for sure that once I have waited several minutes.  And returning and trying another cache listing went into the same infinite wait. 

So there must be some condition, where it gets tripped up so badly, that the only way of recovering is to let Android shut down the app completely and restarting it in flight mode.  Some information cached somewhere ... maybe ... ;-)

**Comment by Lineflyer** [comment]
Just a quick update:
I currently have limited time, but this issue is on my ToDo for analysis and trying to find the root cause.
Up to now I am quite sure its some network thread which is
1. Should not happen at first place as we have all needed data offline to view the cache
2. Not timing out after 30 seconds (as I also experienced a >2 minute waiting time until finally the cache details open

Please give some time as I find time to dive deeper.

**Comment by Lineflyer** [comment]
@DerRadler 
Please kindly report any geocode where this problem appeared to you. Could be extremely helpful.

**Comment by DerRadler** [comment]
Lieflyer, I appreciate. 

 In the next days I will not come across that failure, as I am in Venice and the network is great.  Although a good network was not the main attraction. ;-)  Thanks for taking it!

**Comment by Lineflyer** [comment]
I tried around a lot on a "slowed down" emulator and found no 100% reproduible case, however looking at the network activity while performing different actions I found a scenario, where the "hanging when opening a cache" happens quite often.

Maybe someone can test, confirm and try to analyse deeper (than I can):
- Open c:geo on emulator
- Make sure "Store log images" in c:geo Menu - Offline Data is switched off
- Store some caches (preferable some with many log images, like earth or virtual). Important: Do not open the details yet, but rather store from search result (or similar)
- Shut down emulator
- Throttle emulator (Edit Emulator - Advanced settings - Network speed: EDGE , Network Latency: None)
- Cold boot emulator
- Start c:geo on emulator again
- Go to the stored caches
- Click on one of them, normally it openes directly
- Close it again
- Try to open the same cache again (sometimes also when opening just another cache)
- This mostly leads to "Loading cache details" window for several minutes until the cache details are opened

From what I could see in the debug log (looking at the network traffic) my interpretation (not sure) is:

- On first time opening the cache there are several requests to imgdcn.geocaching.com.... probably to load missing log images. This however happens asynchronous as the cache details are shown immediately
- When closing and opening the cache again I again see many requests to imgcdn.geocaching.com... but this time it seems to block showing the cache details (or the pending network requests from first time opening the cache block it?).
- After 30 seconds I start seeing timeouts for the network requests and following exceptions (not being able to store the file as the socket is closed). The cache details seem to finally open only after the last of those timeouts happened.

**Comment by DerRadler** [comment]
Yes, I believe it is happening when c:geo tries to complete the offline data, most probably with images.

Today I have reproduced the problem with offline access hanging again, and I have collected the details:

The offline cache GC9R3BK was imported together with many others from a GSAK generated GPX file.  I went to the forest into the network-free area.  I have not touched my phone until I was deep into the area with zero network.  Tried to look at cache GC9R3BK.  c:geo got stuck in the process of loading.  I left c:geo, switched into airplane mode and tried again.  Hanging again.  I made Android stop c:geo and tried again in airplane mode:  Success.

So, today it was not happening with a slow network.  It was with a network without the slightest chance of connecting.  Android showed no network at all.

I am happy to share the GPX file with GC9R3BK, as it was before import into c:geo.  I just removed the bulk of other caches, because they might contain corrected coordinates and personal cache notes I would not want to share.
[Ghaus.zip](https://github.com/cgeo/cgeo/files/8886610/Ghaus.zip)

**Comment by Lineflyer** [comment]
I have stumbled upon this problem in my recent vacation a lot. Can someone with code insights help here looking into that problem?? Maybe @moving-bits ?

Let me try to summarize the problem once more:

**Issue Description:**
Clicking on an offline stored cache shows "Loading cache details" for an eternity although the cache should be available offline
Only after waiting several minutes the cache might finally by opened.
This does not happen for all caches but seems to depend on whether log images are present in general and some other yet unknown situation.

**Precondition:**
This happens ONLY if an internet connection is theoretically available but either slow or not working (e.g. non responsive Wifi). As soon as I manually disconnect the internet connection the problem disappears and the cache opens instantly.

**Investigation so far:**
As described in [my comment above](https://github.com/cgeo/cgeo/issues/11725#issuecomment-1126912782) checking on a slow speed emulator clearly shows, that the reason for the delay/blocking is, that c:geo tries to load/refresh images for the cache before opening it (not happening asynchronous).

**Comment by moving-bits** [comment]
Thanks to @Lineflyer's description I'm able to reproduce this behavior in the emulator. Alas, even after some extensive debugging, I have no idea what is blocking currently.

I can see a couple of threads, some for downloading the images. But:

@cgeo/core: How can I find out where the blocking call comes from?

**Comment by Lineflyer** [comment]
I can only contribute my experience of how I think c:geo normally handles those image things:
- On storing a cache c:geo will by default also download all listing images and store them permanently together with its meta-data. Optionally (if enabled in settings) the same applies for log images.
- On accessing non stored images (e.g. log images) c:geo downloads them once the relevant info is needed (e.g. log image or gallery is opened)
- Furthermore on some occasions (not sure under what condition) I know from older logs since years, that c:geo tries to re-download existing images (and usually gets response "304 - Not modified"). 

This all seems pretty normal and usually happens "in the background".

**Comment by DerRadler** [comment]
Happy to see new activitiy here! :-)  Please let me call out two things from the user perspective, just in case it would help maybe:
1. Once the blocking occurs, going into flight mode never helped for me.  I had to explicitly instruct Android to close the app and to remove cached data.  Then going into flight mode and retrying worked.
2. I am using GSAK to prepare GPX files, which then are imported into c:geo.  I assume that something different may happen in this case compared to downloading the cache data directly into c:geo.

Many thanks from a happy user for working on it!
   Frank

**Comment by Lineflyer** [comment]
> I am using GSAK to prepare GPX files, which then are imported into c:geo. I assume that something different may happen in this case compared to downloading the cache data directly into c:geo.

I don't think it makes any difference, however:
Does this problem happen to you once you try to open one of these GPX imported caches (before you ever opened or refreshed them in c:geo after importing)?

**Comment by eddiemuc** [comment]
A thought what maybe happens: 
Afair the number of threads in the pool used to download images is limited. If more images than this limit need download then the call starting the download (in main thread) might block until a thread is freed. If all image downloads.are timing out, one could easily imagine that the whole system blocks at least until the first image download attempts are timing out.

**Comment by Lineflyer** [comment]
> I still don't understand though, why/how this behavior has changed compared to, let's say, a year ago or so? 

IMHO it has not changed in years.
I had this already in my vacation in 2017, but more rare.

That is was more rare back then could however have several reasons outside c:geo:
- No European free data roaming -> device was offline on vacation, thus the problem was not visible
- Less high res pictures on caches/logs

I assume its no recent regression and I might have opened another issue back then (will check).

**Comment by eddiemuc** [comment]
If blocking pool access is the cause of this (and please note that this isnjust a guess of me. I haven't checked in code!), then we could resolve the problem by starting one background thread when cache is opened (from an unlimited pool), and the concrete image load requests are then issued from this background thread as background threads from limited network pool. 

Also I hope that pending http requests are disposed if cache detail activity is closed? And connection request is then immediately terminated?

We should NOT increase the number of threads in network pool though. It is valid to limit c:geo not to issue too many http requests in parallel. Imagine someone opens a cache with e.g. 200 images - cgeo should not open 200 http requests in parallel in this case!

**Comment by DerRadler** [comment]
Again the user speaking up to the persons knowing the code ;-)

> That could explain why the first call works, and only subsequent calls fail. This could be tested by opening and closing the cache, then wait for the timeouts, then reopening the cache. If the theory is correct, that delayed second opening should be as fast as the first.

For me it was always the first attempt to open the cache details, what hung up c:geo for ever (seemingly for ever, but definitely for longer than five minutes).

Trying to load pictures in a task started in background and blocking in background would seem ideal to me.  So opening the cache and showing the details would be quick and would always work, while some pictures may appear late or never.  However, a drawback may be that one is checking if there is a spoiler picture, would see no indication that there is, and the next time looking would find a spoiler picture.  That would be a case to wonder about the own sanity. ;-)

**Comment by Lineflyer** [comment]
@cgeo/team 
Does anyone have the chance to look into this issue again?
With the sequence described in #15477 we have a reproducible testcase to trigger this issue.

We repeatedly get such reports on support mail, where users are unable to open caches in a pure offline scenario (e.g. roaming without data plan). 

I am pretty much sure, that the background image loading activity (see #15477) is the reason for the blockage of the UI in such cases. If we could solve this, it would be great.

**Comment by MagpieFourtyTwo** [comment]
Just stumbled over this issue (thx to @Lineflyer 's reference), read through it, and remembered the problems my cachemate had on our last tour through France.

Background info: Her SIM currently has a problem while roaming, i. e. although network coverage is good, data connection cuts off after a really small amount of data has been transferred or will not be established at all.

Subsequently, she had the described problem a lot.

Scenario: She usually got a GPX in the morning (exported from c:geo), imported it and while still being in our van's WLAN she immediately uploaded changed coordinates and personal notes and finally refreshed the imported caches. Which we thought would be sufficient to be good to go out caching without resp. with "broken" mobile data.

But we were wrong ... she really often got completely stuck on opening a cache ... most regrettably we did not investigate further, cause we just though it's a matter of her SIM ... 

So, as this appears to be a perfect test case, is there any additional info needed or can some special tests be done once we are in roaming area again or is everything clear over here?

Needless to say that we are really interested in solving this issue ... ;)

**Comment by moving-bits** [comment]
I do not have a solution yet for it, but at least I have found some potential culprit:

- I have added some logging to `HTMLImage.fetchDrawableWithMetadata()` and, further down, to `HTMLImage.fetchDrawableUncached()`
- Test case is to open an already stored cache at least twice.
- If there are some pics missing on first run, they will be retrieved by `HTMLImage.fetchDrawableUncached()` => ok
- BUT: If you open the same cache again, _some_ of the pics will be downloaded again (for reasons yet unknown to me)

Some strange observation:
- On retrieving a missing picture, c:geo sends a http header `If-Modified-Since` with some timestamp (different from pic to pic)
- for the pictures that get reloaded every time, the server responds with a `Last-Modified` header with the exact same timestamp as the individual request had as `If-Modified-Since`
